src/compiler/checker.ts

```diff
@@ -4853,6 +4853,13 @@ module ts {
                                 narrowedType = narrowType(type, (<ConditionalExpression>node).condition, /*assumeTrue*/ child === (<ConditionalExpression>node).whenTrue);
                             }
                             break;
+                        case SyntaxKind.CaseClause:
+                            // In a case clause of a switch statement using typeof 
+                            var parent = (<CaseClause>node).parent;
+                            if((<SwitchStatement>parent).expression.kind === SyntaxKind.TypeOfExpression){

```

Comment:
Make sure the parent is guaranteed to be a switch statement. It is possible that the parser lets through case labels that are not parented by a switch. Please add either an assert or a bailout condition.